base/channels.jl

```diff
@@ -166,9 +166,13 @@ isbuffered(c::Channel) = c.sz_max==0 ? false : true
 
 function check_channel_state(c::Channel)
     if !isopen(c)
-        excp = c.excp
-        excp !== nothing && throw(excp)
-        throw(closed_exception())
+        lock(c)
+        excp = try
+            @something(c.excp, closed_exception())
+        finally
+            unlock(c)
+        end
+        throw(excp)
```

Comment:
perhaps this instead would be simpler?
```suggestion
        @assert ((@atomic :acquire c.state) !== :open)
        excp = c.excp
        excp !== nothing && throw(excp)
        throw(closed_exception())
 ```